Show toolbox & distrobox containers different in containers listing

subhoghoshX opened the pull request:
fixes #865

![image](https://github.com/cockpit-project/cockpit-podman/assets/108616679/57766555-f315-4035-99f4-d7ce24799ba7)

---

# Show toolbox & distrobox containers different in containers listing

![image](https://github.com/cockpit-project/cockpit-podman/assets/67428/36cc4aec-a90d-4930-b9ac-7e0e82a8c222)

subhoghoshX commented:
I wonder whether gray looks better:

![image](https://github.com/cockpit-project/cockpit-podman/assets/108616679/7e961d01-4d37-4445-81fe-7ee416b75085)

cockpituous commented on src/Containers.jsx:
```diff
@@ -397,7 +398,10 @@ class Containers extends React.Component {
-                <span className="container-name">{container.Name}</span>
+                <Flex alignItems={{ default: 'alignItemsCenter' }}>
+                    <span className="container-name">{container.Name}</span>
+                    {isToolboxContainer && <Badge>toolbox</Badge>}
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1409-20230914-111755-35d42e9d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1409-20230915-112344-e0e65ca9-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

garrett commented:
Yeah, blue is probably too in-your-face. Grey works, but might be too understated. I wonder if any other color could work. (Green, orange, red are all probably out, as they often mean "good", "problem", and "error".)

Perhaps a light purple or light cyan?
https://www.patternfly.org/design-foundations/colors#color-palette

Can you also add distrobox support? It should be the same, but say "distrobox" instead of "toolbox". Thanks!

jelly reviewed (commented):
LGTM, but needs a design review from @garrett

jelly commented:
> Yeah, blue is probably too in-your-face. Grey works, but might be too understated. I wonder if any other color could work. (Green, orange, red are all probably out, as they often mean "good", "problem", and "error".)
> 
> Perhaps a light purple or light cyan? https://www.patternfly.org/design-foundations/colors#color-palette
> 
> Can you also add distrobox support? It should be the same, but say "distrobox" instead of "toolbox". Thanks!

Seems it uses a different label? https://github.com/89luca89/distrobox/blob/3099d53a71ea5511eec23e83c33e25c032729e4f/distrobox-create#L579

jelly commented:
A test would be great, just create a container with a label which matches toolbox's behaviour

subhoghoshX commented:
Now it looks like this (I went with purple for toolbox and gold for distrobox, which is very loosely based on their brand colors):

![image](https://github.com/cockpit-project/cockpit-podman/assets/108616679/bb8526fc-aa4e-423d-885c-80e66f8b6d4d)

garrett commented:
Previously, I was thinking that we'd just use one color, because how many people use both toolbox _and_ distrobox? I mean, I do, and perhaps others on the team as well... but we're probably not that ordinary. :wink: 

But this works for me.

My one concern is the contrast ratio for the colors, primarily the gold color.

If I plug the colors into https://accessible-colors.com/ — I get this:

![image](https://github.com/cockpit-project/cockpit-podman/assets/10246/8db8851c-3f71-438d-9c93-99b6eeeb92fb)

Which means we either need to make the background or foreground darker to have the correct color contrast ratio, so people can more easily read it.

The purple also has the same problem, but not as much:

![image](https://github.com/cockpit-project/cockpit-podman/assets/10246/dc936602-4b39-48af-ad13-762b982f2646)

We'll want to stick with the PF color palette (extended), so moving a shade down for the background colors of each would work... Or moving the shade up and flipping the text to dark instead of light.

(At the size and weight do matter, but not at this size or smaller, which this is. For the purple to pass without color changes, it'd have to be 18px and bold... which we definitely don't want to do here.)

https://www.patternfly.org/design-foundations/colors/#color-palette

subhoghoshX commented:
Fixed the contrast issues @garrett, new look:

![image](https://github.com/cockpit-project/cockpit-podman/assets/108616679/188f843d-5e62-4d66-9c9c-df2c35450050)

In dark mode it looks good:

![image](https://github.com/cockpit-project/cockpit-podman/assets/108616679/44c0f985-518f-41da-a864-b595dfbc02c4)

PS: Using https://accessible-colors.com/ the ![image](https://github.com/cockpit-project/cockpit-podman/assets/108616679/3e66c0cc-67d6-4eb4-be3b-352fc56e1b1b) badge doesn't seem to have a good contrast.

subhoghoshX commented:
umm not sure why the tests failed...

garrett commented:
Ah! I was working on color choices and used the hex on that website... and made this:

![image](https://github.com/cockpit-project/cockpit-podman/assets/10246/ebeeda7f-884e-47ec-94dd-a09daf3e892f)

subhoghoshX commented:
> Ah! I was working on color choices and used the hex on that website... and made this:
> 
> ![image](https://user-images.githubusercontent.com/10246/269670301-ebeeda7f-884e-47ec-94dd-a09daf3e892f.png)

Thanks @garrett. So which one should we go for? Or use both? second one for light theme and first one for dark theme?

garrett commented:
I like the first set overall (dark on light), but if we have to do custom CSS anyway for assigning the colors, then we could do the light/dark, with the first set as light and second set as dark.

subhoghoshX commented:
Thanks @garrett, it indeed looks cool now.

Light:

![image](https://github.com/cockpit-project/cockpit-podman/assets/108616679/74513844-8470-4238-a1b3-02f4af9e1b3d)

Dark:

![image](https://github.com/cockpit-project/cockpit-podman/assets/108616679/9320de42-3d0a-4e8c-bf9d-91aa1f107583)

garrett commented on src/podman.scss:
```diff
@@ -69,6 +69,26 @@
   color: white;
 }
 
+.ct-badge-toolbox {
+    background-color: var(--pf-v5-global--palette--purple-100);
+    color: var(--pf-v5-global--palette--purple-600);
+}
+
+.ct-badge-distrobox {
+    background-color: var(--pf-v5-global--palette--gold-100);
+    color: var(--pf-v5-global--palette--gold-600);
+}
+
+.pf-v5-theme-dark .ct-badge-toolbox {
+    background-color: var(--pf-v5-global--palette--purple-500);
+    color: white;
+}
+
+.pf-v5-theme-dark .ct-badge-distrobox {
+    background-color: var(--pf-v5-global--palette--gold-500);
+    color: white;
+}
```
As this is SCSS instead of CSS, it's also possible to write it with nesting, so the associated CSS values are grouped by type, like this:

```scss
.ct-badge-toolbox {
    background-color: var(--pf-v5-global--palette--purple-100);
    color: var(--pf-v5-global--palette--purple-600);

    .pf-v5-theme-dark & {
        background-color: var(--pf-v5-global--palette--purple-500);
        color: white;
    }
}

.ct-badge-distrobox {
    background-color: var(--pf-v5-global--palette--gold-100);
    color: var(--pf-v5-global--palette--gold-600);

    .pf-v5-theme-dark & {
        background-color: var(--pf-v5-global--palette--gold-500);
        color: white;
    }
}
```

This is also the way you'd write plain CSS with nesting, which is a new feature inspired by SCSS. https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting — but while native nesting is now available in all browsers (except Firefox extended support release), we can't use it yet. But we're using SCSS, so it's fine to use where we know it'll get compiled down. https://caniuse.com/css-nesting

It's not a big deal though; what's here in the PR is totally correct. :grin:

garrett reviewed (approved):
:+1: Looks great! Thanks for working on this.

I've added a note about an alternate way to write the SCSS (which is probably how I would've done it), but it's totally fine as-is.

subhoghoshX commented:
> 👍 Looks great! Thanks for working on this.
> 
> I've added a note about an alternate way to write the SCSS (which is probably how I would've done it), but it's totally fine as-is.

Makes sense. It did cross my mind once to nest it, but I brushed it off for some reason :P... fixed it! thanks again!

jelly reviewed (approved):
Thanks
